### PR TITLE
Migrated the Scala driver to the MongoDB reactive streams driver

### PR DESCRIPTION
**Pull request by rozza**
SCALA-556

### CODE REVIEW

File: docs/reference/content/getting-started/quick-tour.md [hunk 1]
```diff
@@ -129,7 +129,7 @@ In the API all methods returning a `Observables` are "cold" streams meaning that
 The example below does nothing:
 
 ```scala
-val observable: Observable[Completed] = collection.insertOne(doc)
+val observable: SingleObservable[Completed] = collection.insertOne(doc)
```
Comment (rozza) [inline docs/reference/content/getting-started/quick-tour.md:132]: SingleObservable is more correct than Observable for inserts.

File: driver/src/it/scala/org/mongodb/scala/DocumentationTransactionsExampleSpec.scala [hunk 2]
```diff
@@ -43,8 +43,8 @@ class DocumentationTransactionsExampleSpec extends RequiresMongoDBISpec {
     client.getDatabase("hr").createCollection("employees").execute()
     client.getDatabase("hr").createCollection("events").execute()
 
-    updateEmployeeInfoWithRetry(client).execute() should equal(Completed())
-    client.getDatabase("hr").drop().execute() should equal(Completed())
+    updateEmployeeInfoWithRetry(client).execute()
```
Comment (rozza) [inline driver/src/it/scala/org/mongodb/scala/DocumentationTransactionsExampleSpec.scala:46]: `Completed()` has been removed and `Completed` is now just a type alias pointing to `Void`.

This means all execute wont return anything so `should equal(Completed())` can be removed

File: driver/src/main/scala/org/mongodb/scala/AggregateObservable.scala [hunk 3]
```diff
@@ -164,7 +163,15 @@ case class AggregateObservable[TResult](private val wrapped: AggregateIterable[T
...
-  def toCollection(): Observable[Completed] = observeCompleted(wrapped.toCollection(_: SingleResultCallback[Void]))
+  def toCollection(): SingleObservable[Completed] = wrapped.toCollection()
+
+  /**
+   * Helper to return a single observable limited to the first result.
+   *
+   * @return a single observable which will the first result.
+   * @since 4.0
+   */
+  def first(): SingleObservable[TResult] = wrapped.first()
```
Comment (rozza) [inline driver/src/main/scala/org/mongodb/scala/AggregateObservable.scala:174]: Looks like lots of Observables didn't have a `first()` method, probably because the scala extension to `Observable` added a `head` method.  Added to mirror the reactive streams api.

File: driver/src/main/scala/org/mongodb/scala/ChangeStreamObservable.scala [hunk 4]
```diff
@@ -140,7 +139,15 @@ case class ChangeStreamObservable[TResult](private val wrapped: ChangeStreamIter
    * @tparam T the result type
    * @return an Observable
    */
-  def withDocumentClass[T](clazz: Class[T]): Observable[T] = observe(wrapped.withDocumentClass(clazz))
+  def withDocumentClass[T](clazz: Class[T]): Observable[T] = wrapped.withDocumentClass(clazz).toObservable()
```
Comment (rozza) [inline driver/src/main/scala/org/mongodb/scala/ChangeStreamObservable.scala:142]: Had to help the compiler here, so `Publisher[T]` can be implicitly converted to `Observable[T]`. 

This is the only place I've had to explicitly use the `toObservable()` helper. That in itself is odd as the `toObservable` method is an implicit method on the implicit conversion class that converts `Publisher[T]` to `Observable[T]`.

File: driver/src/main/scala/org/mongodb/scala/MongoClient.scala [hunk 5]
```diff
@@ -66,33 +64,6 @@ object MongoClient {
       .build(), mongoDriverInformation)
   }
 
-  /**
```
Comment (rozza) [inline driver/src/main/scala/org/mongodb/scala/MongoClient.scala:69]: I removed these but this can probably be split out in a separate commit.

File: driver/src/main/scala/org/mongodb/scala/Observable.scala [hunk 6]
```diff
@@ -39,14 +42,34 @@ object Observable {
 }
 
 /**
- * A `Observable` represents a MongoDB operation.
+ * A `Observable` represents a MongoDB operation and implements the `Publisher` interface.
```
Comment (rozza) [inline driver/src/main/scala/org/mongodb/scala/Observable.scala:45]: This file contains the big change. Previously we has the async `Observable` and a scala implicit `Observable` that would add all these helpers.

Now they are part of the trait and the `Observable` itself implements the `Publisher` interface.

File: driver/src/main/scala/org/mongodb/scala/ObservableImplicits.scala [hunk 7]
```diff
@@ -16,420 +16,74 @@
...
+import org.mongodb.scala.bson.ObjectId
+import org.mongodb.scala.gridfs.GridFSFile
 import org.mongodb.scala.internal._
+import org.reactivestreams.{Publisher, Subscriber, Subscription => JSubscription}
+
+import scala.concurrent.Future
 
 /**
- * Extends the Java [[Observable]] and adds helpers to make Observables composable and simple to Subscribe to.
+ * Implicit conversion support for Publishers, Observables and Subscriptions
```
Comment (rozza) [inline driver/src/main/scala/org/mongodb/scala/ObservableImplicits.scala:30]: Most of ObservableImplicits moved into the `Observable` trait.

ObservableImplicits now helps with the translation between raw `Publisher` to scala `Observable`. There is the `SingleObservable` converter as well.

There are also implicit class helpers for type conversion so `Publisher[java.lang.Long]` can become the scala equivalent: `Observable[Long]`

File: driver/src/main/scala/org/mongodb/scala/Observer.scala [hunk 8]
```diff
@@ -35,7 +35,7 @@ import com.mongodb.async.client.{Observer => JObserver, Subscription => JSubscri
  *
  * @tparam T The type of element signaled.
  */
-trait Observer[T] extends JObserver[T] {
+trait Observer[T] extends Subscriber[T] {
```
Comment (rozza) [inline driver/src/main/scala/org/mongodb/scala/Observer.scala:38]: `Observer[T`] extends `Subscriber[T]` the only difference is it can take either a reactive streams `Subscription` or a scala driver `Subscription`.

File: driver/src/main/scala/org/mongodb/scala/Subscription.scala [hunk 9]
```diff
@@ -55,4 +55,11 @@ trait Subscription extends JSubscription {
    * @return `true` if this `Subscription` is currently unsubscribed, `false` otherwise
    */
   def isUnsubscribed: Boolean
+
+  /**
+   * Request the [[Observable]] to stop sending data and clean up resources.
+   *
+   * As this request is asynchronous data may still be sent to meet previously signalled demand after calling cancel.
+   */
+  override def cancel(): Unit = unsubscribe()
```
Comment (rozza) [inline driver/src/main/scala/org/mongodb/scala/Subscription.scala:64]: The scala `Subscription` is very slightly different to the reactivestreams one - it has an `unsubscribe` method rather than a `cancel` method, it also has a `isUnsubscribed` method.

File: driver/src/main/scala/org/mongodb/scala/gridfs/AsyncInputStream.scala [hunk 10]
```diff
@@ -35,7 +35,7 @@ trait AsyncInputStream {
    * @return an Observable with a single element indicating total number of bytes read into the buffer, or
    *         `-1` if there is no more data because the end of the stream has been reached.
    */
-  def read(dst: ByteBuffer): Observable[Int]
+  def read(dst: ByteBuffer): SingleObservable[Int]
```
Comment (rozza) [inline driver/src/main/scala/org/mongodb/scala/gridfs/AsyncInputStream.scala:38]: The SingleObservable type is the correct type.

File: driver/src/main/scala/org/mongodb/scala/gridfs/helpers/AsynchronousChannelHelper.scala [hunk 11]
```diff
@@ -19,7 +19,7 @@ package org.mongodb.scala.gridfs.helpers
 import java.nio.channels.{AsynchronousByteChannel, AsynchronousFileChannel}
 
 import com.mongodb.async.client.gridfs.helpers.{AsynchronousChannelHelper => JAsynchronousChannelHelper}
```
Comment (rozza) [inline driver/src/main/scala/org/mongodb/scala/gridfs/helpers/AsynchronousChannelHelper.scala:21]: I added a ticket to get the AsynchronousChannelHelper ported to reactivestreams.

File: driver/src/test/scala/org/mongodb/scala/AggregateObservableSpec.scala [hunk 12]
```diff
@@ -18,36 +18,27 @@ package org.mongodb.scala
...
-
+import com.mongodb.reactivestreams.client.AggregatePublisher
 import org.mongodb.scala.model.Collation
 import org.scalamock.scalatest.proxy.MockFactory
 import org.scalatest.{FlatSpec, Matchers}
 
-class AggregateObservableSpec extends FlatSpec with Matchers with MockFactory {
+import scala.concurrent.duration.Duration
 
-  def observer[T]: Observer[T] = new Observer[T]() {
```
Comment (rozza) [inline driver/src/test/scala/org/mongodb/scala/AggregateObservableSpec.scala:31]: For the tests we no longer need an Observable we can just test the publishers method is called.

**Review by rozza — COMMENTED** [review]
Patch: https://evergreen.mongodb.com/version/5d713a11c9ec447de5782e8c

File: docs/reference/content/getting-started/quick-tour.md (hunk 1 above)
Comment (jyemin) [reply docs/reference/content/getting-started/quick-tour.md:132]: Is this a breaking change that requires application source code modifications?

File: driver/src/it/scala/org/mongodb/scala/DocumentationTransactionsExampleSpec.scala (hunk 2 above)
Comment (jyemin) [reply driver/src/it/scala/org/mongodb/scala/DocumentationTransactionsExampleSpec.scala:46]: Is this also a breaking change requiring source code modifications?

File: driver/src/main/scala/org/mongodb/scala/AggregateObservable.scala (hunk 3 above)
Comment (jyemin) [reply driver/src/main/scala/org/mongodb/scala/AggregateObservable.scala:174]: ack

File: docs/reference/content/getting-started/quick-tour.md (hunk 1 above)
Comment (jyemin) [reply docs/reference/content/getting-started/quick-tour.md:132]: Maybe not, since SingleObservable extends Observable?

File: driver/src/main/scala/org/mongodb/scala/ChangeStreamObservable.scala (hunk 4 above)
Comment (jyemin) [reply driver/src/main/scala/org/mongodb/scala/ChangeStreamObservable.scala:142]: Is it important to understand why?  I'm not sure.

File: driver/src/main/scala/org/mongodb/scala/ClientSessionImplicits.scala [hunk 13]
```diff
@@ -37,13 +35,14 @@ trait ClientSessionImplicits {
      *
      * A transaction can only be commmited if one has first been started.
      */
-    def commitTransaction(): SingleObservable[Completed] = observeCompleted(clientSession.commitTransaction(_: SingleResultCallback[Void]))
+    def commitTransaction(): SingleObservable[Completed] = clientSession.commitTransaction()
```
Comment (jyemin) [inline driver/src/main/scala/org/mongodb/scala/ClientSessionImplicits.scala:38]: Nice and simple now!

File: driver/src/main/scala/org/mongodb/scala/MongoClient.scala (hunk 5 above)
Comment (jyemin) [reply driver/src/main/scala/org/mongodb/scala/MongoClient.scala:69]: It would make the commit history clearer, but up to you.

File: driver/src/main/scala/org/mongodb/scala/MongoCollection.scala [hunk 14]
```diff
@@ -130,7 +129,7 @@ case class MongoCollection[TResult](private val wrapped: JMongoCollection[TResul
    * @return a Observable with a single element indicating the number of documents
    */
   @deprecated("Use `countDocuments` or `estimatedDocumentCount()` instead of `count()`", "2.4")
-  def count(): SingleObservable[Long] = observeLong(wrapped.count(_: SingleResultCallback[java.lang.Long]))
+  def count(): SingleObservable[Long] = wrapped.count()
```
Comment (jyemin) [inline driver/src/main/scala/org/mongodb/scala/MongoCollection.scala:132]: Remind me how `wrapped` gets initialized.

File: driver/src/main/scala/org/mongodb/scala/Observable.scala [hunk 15]
```diff
@@ -66,5 +89,333 @@ trait Observable[T] extends JObservable[T] {
    * @note Users should not have to implement this method but rather use the Scala `Observable`.
    * @param observer the `Observer` that will consume signals from this `Observable`
    */
-  override def subscribe(observer: JObserver[_ >: T]): Unit = subscribe(BoxedObserver[T](observer))
+  override def subscribe(observer: Subscriber[_ >: T]): Unit = subscribe(BoxedSubscriber[T](observer))
```
Comment (jyemin) [inline driver/src/main/scala/org/mongodb/scala/Observable.scala:92]: This change will require application source code changes, right?

File: driver/src/main/scala/org/mongodb/scala/ObservableImplicits.scala (hunk 7 above)
Comment (jyemin) [reply driver/src/main/scala/org/mongodb/scala/ObservableImplicits.scala:30]: I need a walkthrough if this, I think.

File: driver/src/main/scala/org/mongodb/scala/Observer.scala (hunk 8 above)
Comment (jyemin) [reply driver/src/main/scala/org/mongodb/scala/Observer.scala:38]: Because there's an implicit conversion from one to the other?

File: driver/src/main/scala/org/mongodb/scala/Subscription.scala (hunk 9 above)
Comment (jyemin) [reply driver/src/main/scala/org/mongodb/scala/Subscription.scala:64]: ack

File: driver/src/main/scala/org/mongodb/scala/gridfs/helpers/AsynchronousChannelHelper.scala (hunk 11 above)
Comment (jyemin) [reply driver/src/main/scala/org/mongodb/scala/gridfs/helpers/AsynchronousChannelHelper.scala:21]: ack

File: driver/src/main/scala/org/mongodb/scala/gridfs/package.scala [hunk 16]
```diff
@@ -49,69 +49,31 @@ package object gridfs {
...
-    )
-    // scalastyle:on null
-  }
-
-  implicit class ScalaAsyncOutputStreamToJava(wrapped: AsyncOutputStream) extends JAsyncOutputStream {
-    // scalastyle:off null
-    override def write(src: ByteBuffer, callback: SingleResultCallback[Integer]): Unit = wrapped.write(src).subscribe(
-      new Observer[Int] {
-        var bytesWritten: Option[Int] = None
+    override def read(dst: ByteBuffer): Publisher[java.lang.Integer] = wrapped.read(dst).map(i => new java.lang.Integer(i))
```
Comment (jyemin) [inline driver/src/main/scala/org/mongodb/scala/gridfs/package.scala:67]: I'm missing something here.  Can you explain why this can be so much simpler now? :)

File: driver/src/main/scala/org/mongodb/scala/package.scala [hunk 17]
```diff
@@ -358,6 +358,13 @@ package object scala extends ClientSessionImplicits with ObservableImplicits wit
    */
   type ClientEncryptionSettings = com.mongodb.ClientEncryptionSettings
 
+  /**
+   * Alias to Void
+   *
+   * @since 4.0
+   */
+  type Completed = Void
```
Comment (jyemin) [inline driver/src/main/scala/org/mongodb/scala/package.scala:366]: Where was this type defined previously?

File: driver/src/test/scala/org/mongodb/scala/model/vault/ClientEncryptionSpec.scala [hunk 18]
```diff
@@ -48,23 +41,27 @@ class ClientEncryptionSpec extends FlatSpec with Matchers with MockFactory {
...
-    wrapped.expects(Symbol("encrypt"))(bsonValue, options, *).once()
+    wrapped.expects(Symbol("encrypt"))(bsonValue, options).once()
 
-    clientEncryption.encrypt(bsonValue, options).subscribe(observer[BsonBinary])
+    clientEncryption.encrypt(bsonValue, options)
   }
 
   it should "call decrypt" in {
     val bsonBinary = BsonBinary(Array[Byte](1, 2, 3))
-    wrapped.expects(Symbol("decrypt"))(bsonBinary, *).once()
```
Comment (jyemin) [inline driver/src/test/scala/org/mongodb/scala/model/vault/ClientEncryptionSpec.scala:65]: What did the `*` mean?

**Review by jyemin — COMMENTED** [review]
Looks really nice, Ross.  

I don't think I have any change requests, just a bunch of questions.  In particular, I'd like to understand a little bit better what (if any) source modifications will be required for upgraders.

File: driver/src/main/scala/org/mongodb/scala/Observer.scala (hunk 8 above)
Comment (rozza) [reply driver/src/main/scala/org/mongodb/scala/Observer.scala:38]: In this case it needs to be able to take either due to the slightly different api of Subscription between the original implementation and reactive streams.

File: docs/reference/content/getting-started/quick-tour.md (hunk 1 above)
Comment (rozza) [reply docs/reference/content/getting-started/quick-tour.md:132]: Correct

File: driver/src/it/scala/org/mongodb/scala/DocumentationTransactionsExampleSpec.scala (hunk 2 above)
Comment (rozza) [reply driver/src/it/scala/org/mongodb/scala/DocumentationTransactionsExampleSpec.scala:46]: Minor chance of a break here.

If the user relied on `onNext()` being called with a Completed Token then that would have to be updated to use `onComplete()` instead.  This change is akin to the change requested in JAVA-3398

File: driver/src/main/scala/org/mongodb/scala/ChangeStreamObservable.scala (hunk 4 above)
Comment (rozza) [reply driver/src/main/scala/org/mongodb/scala/ChangeStreamObservable.scala:142]: I think this is more of a Scala implicit resolution issue - I tried to work around it but failed to find the solution in this case.

File: driver/src/main/scala/org/mongodb/scala/gridfs/package.scala (hunk 16 above)
Comment (rozza) [reply driver/src/main/scala/org/mongodb/scala/gridfs/package.scala:67]: Because all we have to do now is wrap the publisher that does all this underneath and convert from the Scala Int type to the java type.

File: driver/src/main/scala/org/mongodb/scala/package.scala (hunk 17 above)
Comment (rozza) [reply driver/src/main/scala/org/mongodb/scala/package.scala:366]: It was an Object that used to live here: driver/src/main/scala/org/mongodb/scala/Completed.scala

File: driver/src/test/scala/org/mongodb/scala/model/vault/ClientEncryptionSpec.scala (hunk 18 above)
Comment (rozza) [reply driver/src/test/scala/org/mongodb/scala/model/vault/ClientEncryptionSpec.scala:65]: `*` means match any parameter, we used it match the callbacks.

File: driver/src/main/scala/org/mongodb/scala/ObservableImplicits.scala [hunk 19]
```diff
@@ -16,420 +16,74 @@
...
-    /**
-     * Use a specific execution context for future operations
-     *
-     * @param context the execution context
-     * @return an Observable that uses the specified execution context
-     */
-    def observeOn(context: ExecutionContext): Observable[T] = ExecutionContextObservable(observable, context)
+  }
 
+  implicit class ToObservableString(publisher: Publisher[java.lang.String]) extends Observable[String] {
```
Comment (jyemin) [inline driver/src/main/scala/org/mongodb/scala/ObservableImplicits.scala:72]: Nit: Maybe name it `ToObservableOfString`

File: driver/src/main/scala/org/mongodb/scala/ChangeStreamObservable.scala (hunk 4 above)
Comment (jyemin) [reply driver/src/main/scala/org/mongodb/scala/ChangeStreamObservable.scala:142]: ack

File: driver/src/main/scala/org/mongodb/scala/MongoCollection.scala (hunk 14 above)
Comment (jyemin) [reply driver/src/main/scala/org/mongodb/scala/MongoCollection.scala:132]: Never mind, I see now.

File: driver/src/main/scala/org/mongodb/scala/Observable.scala (hunk 15 above)
Comment (jyemin) [reply driver/src/main/scala/org/mongodb/scala/Observable.scala:92]: Ping on this one.  I'm still not clear how the implicits will solve this.

File: driver/src/main/scala/org/mongodb/scala/ObservableImplicits.scala (hunk 7 above)
Comment (jyemin) [reply driver/src/main/scala/org/mongodb/scala/ObservableImplicits.scala:30]: Got the walkthrough.  Thanks!

File: driver/src/test/scala/org/mongodb/scala/AggregateObservableSpec.scala (hunk 12 above)
Comment (jyemin) [reply driver/src/test/scala/org/mongodb/scala/AggregateObservableSpec.scala:31]: ack
